### PR TITLE
add proof output for normal, packed or explicit

### PR DESCRIPTION
**Pull request by digama0**
Compressed proof output on the way

### CODE REVIEW

File: src/diag.rs [hunk 1]
```diff
@@ -187,7 +187,7 @@ fn annotate_diagnostic(notes: &mut Vec<Notation>,
     }
 
     fn t(v: &Token) -> String {
-        String::from_utf8((**v).to_owned()).expect("utf-8 is checked before making Token")
+        unsafe { String::from_utf8_unchecked((**v).to_owned()) }
```
Comment (sorear) [inline src/diag.rs:190]: Why this change?  Currently all of our unsafe blocks 1. locally check their preconditions 2. live in utils.rs 3. have a measurable performance impact; I'm especially concerned by the fact that this has a nonlocal dependency

Comment (digama0) [inline src/diag.rs:190]: Moving it to utils is also reasonable. I added this for consistency with `parser::as_str`. I've been vaguely worried about the accumulation of unnecessary string checks when converting `Vec<[u8]>` to `String` and `&[u8]` to `&str`. The rust people ensure safety by maintaining a correctness contract with the type, and we could do the same thing with `Token` / `TokenPtr` if they were newtypes with checked `new()`. We're obviously maintaining our own invariants, so this call itself is perfectly safe; the issue is with how tokens get allocated in the first place.

Comment (sorear) [inline src/diag.rs:190]: Most `Token` and `TokenPtr` which are constructed never get converted to `&str`, though, so checking should be deferred to conversion time IMO.

I guess I don't understand the motivation?  It can't be a runtime optimization, since that function is called zero times in a typical run; is it trying to shave bytes off the binary?

Comment (digama0) [inline src/diag.rs:190]: I thought all the `Token` objects are created during parsing, and we certainly are checking for ASCII there. Anything which comes from the file buffer is checked ASCII (although it may be hard to make the checking happen at exactly `new()` without a performance hit, so maybe the parser gets a pass on the checking via some `new_unchecked()`).

Currently, as I said this is only changed for consistency with `as_str`. You should focus on that, which does get called a lot. In particular, exporting a large proof will cause many many calls to `as_str` (although since I do lookup beforehand, it's proportional to the number of compressed steps, not uncompressed, when outputting in `Normal` style).

This function should probably be an `as_string` function in `parser` or `utils`, but since you made a local function I assumed you were using it a lot here so I didn't try to factor it out.

File: src/parser.rs [hunk 2]
```diff
@@ -333,6 +333,12 @@ pub fn copy_token(ptr: TokenPtr) -> Token {
     ptr.to_owned().into_boxed_slice()
 }
 
+/// Transmutes this token into a Rust string. This is safe for us
+/// because we check that our tokens are valid ASCII (and thus also valid UTF8).
+pub fn as_str(ptr: TokenPtr) -> &str {
+    unsafe { str::from_utf8_unchecked(ptr) }
```
Comment (sorear) [inline src/parser.rs:339]: Do you have benchmarks showing that the utf-8 check consumes more than 1% of runtime under a plausible workload?  I think the burden of proof should be on whoever is adding a new unsafe block, _especially_ if that unsafe block doesn't have a clear abstraction barrier and effectively depends on everything else in the program for correctness.

Comment (digama0) [inline src/parser.rs:339]: TBH I'm not clear on why this is `unsafe` in the first place. Generally type contract violations do not cause memory unsafety, but I guess this one does? Anyway, there _is_ a clear abstraction barrier: `Token` and `TokenPtr` are required by contract to store only valid ASCII characters. With this abstraction, `as_str` becomes safe, but `new()` becomes unsafe (or more accurately, coercion from `&[u8]` is unsafe).

This needs a separate PR, but a full encapsulation of `Token` and `TokenPtr`, with no way to create these beyond private methods, should solve the issue.
